module/icp/algs/aes/aes_impl.c

```diff
@@ -1567,52 +211,214 @@ aes_alloc_keysched(size_t *size, int kmflag)
 	return (NULL);
 }
 
+/* AES implementation that contains the fastest methods */
+static aes_impl_ops_t aes_fastest_impl = {
+	.name = "fastest"
+};
+
+/* All compiled in implementations */
+const aes_impl_ops_t *aes_all_impl[] = {
+	&aes_generic_impl,
+#if defined(__x86_64)
+	&aes_x86_64_impl,
+#endif
+#if defined(__x86_64) && defined(HAVE_AES)
+	&aes_aesni_impl,
+#endif
+};
+
+/* Indicate that benchmark has been completed */
+static boolean_t aes_impl_initialized = B_FALSE;
+
+/* Select aes implementation */
+#define	IMPL_FASTEST	(UINT32_MAX)
+#define	IMPL_CYCLE	(UINT32_MAX-1)
 
-#ifdef __amd64
+#define	AES_IMPL_READ(i) (*(volatile uint32_t *) &(i))
 
-#define	INTEL_AESNI_FLAG (1 << 25)
+static uint32_t icp_aes_impl = IMPL_FASTEST;
+static uint32_t user_sel_impl = IMPL_FASTEST;
+
+/* Hold all supported implementations */
+static size_t aes_supp_impl_cnt = 0;
+static aes_impl_ops_t *aes_supp_impl[ARRAY_SIZE(aes_all_impl)];
 
 /*
- * Return 1 if executing on Intel with AES-NI instructions,
- * otherwise 0 (i.e., Intel without AES-NI or AMD64).
- * Cache the result, as the CPU can't change.
+ * Selects the aes operations for encrypt/decrypt/key setup
  */
-static int
-intel_aes_instructions_present(void)
+aes_impl_ops_t *
+aes_impl_get_ops()
 {
-	static int cached_result = -1;
-	unsigned eax, ebx, ecx, edx;
-	unsigned func, subfunc;
+	aes_impl_ops_t *ops = NULL;
+	const uint32_t impl = AES_IMPL_READ(icp_aes_impl);
 
-	if (cached_result == -1) { /* first time */
-		/* check for an intel cpu */
-		func = 0;
-		subfunc = 0;
+	switch (impl) {
+	case IMPL_FASTEST:
+		ASSERT(aes_impl_initialized);
+		ops = &aes_fastest_impl;
+		break;
+	case IMPL_CYCLE:
+	{
+		ASSERT(aes_impl_initialized);
+		ASSERT3U(aes_supp_impl_cnt, >, 0);
+		/* Cycle through supported implementations */
+		static size_t cycle_impl_idx = 0;
+		size_t idx = (++cycle_impl_idx) % aes_supp_impl_cnt;
+		ops = aes_supp_impl[idx];
+	}
+	break;
+	default:
+		ASSERT3U(impl, <, aes_supp_impl_cnt);
+		ASSERT3U(aes_supp_impl_cnt, >, 0);
+		if (impl < ARRAY_SIZE(aes_all_impl))
+			ops = aes_supp_impl[impl];
```

Comment:
This assert is there because ARRAY_SIZE(aes_all_impl) is probably greater than aes_supp_impl_cnt.  Hypothetically if some strange error occurred while setting the implementation, impl could be greater than aes_supp_impl_cnt but less than ARRAY_SIZE(aes_all_impl), causing you to get a bogus pointer (in c99 static fixed size arrays (i.e. aes_all_impl) are zeroed, but I have no idea for the gnu89 spec the kernel uses).

If we wanted to remove an assertion, I could change the if statement to use aes_supp_impl_cnt instead of ARRAY_SIZE(aes_all_impl), as aes_supp_impl_cnt will always be less than ARRAY_SIZE(aes_all_impl) save for some cosmic ray bit flipping.

Reply:
ok. in that case its probably fine as it is.